kernel/src/log_segment.rs

```diff
@@ -57,126 +70,93 @@ pub(crate) struct LogSegment {
-            ))
-        );
-
-        // Commit file versions must be greater than the most recent checkpoint version if it exists
-        let checkpoint_version = checkpoint_parts.first().map(|checkpoint_file| {
-            ascending_commit_files.retain(|log_path| checkpoint_file.version < log_path.version);
-            checkpoint_file.version
-        });
-
-        // There must be no gap between a checkpoint and the first commit version. Note that
-        // that all checkpoint parts share the same version.
-        if let (Some(checkpoint_version), Some(commit_file)) =
-            (checkpoint_version, ascending_commit_files.first())
-        {
-            require!(
-                checkpoint_version + 1 == commit_file.version,
-                Error::InvalidCheckpoint(format!(
-                    "Gap between checkpoint version {} and next commit {}",
-                    checkpoint_version, commit_file.version,
-                ))
-            )
-        }
+/// A builder for constructing a [`LogSegment`]. One can specify the start/end versions, etc. and
+/// then call [`LogSegmentBuilder::build_complete`] or [`LogSegmentBuilder::build_incremental`] to
+/// create a [`LogSegment`].
+///
+/// Complete log segments include all actions for current table state (used for snapshot/scan/etc.)
+/// whereas incremental log segments include all actions between two versions. End version is
+/// inclusive.
+#[internal_api]
+pub(crate) struct LogSegmentBuilder {
+    log_root: Url,
+    end_version: Option<Version>,
+}
 
-        // Get the effective version from chosen files
-        let effective_version = ascending_commit_files
-            .last()
-            .or(checkpoint_parts.first())
-            .ok_or(Error::generic("No files in log segment"))?
-            .version;
-        if let Some(end_version) = end_version {
-            require!(
-                effective_version == end_version,
-                Error::generic(format!(
-                    "LogSegment end version {effective_version} not the same as the specified end version {end_version}"
-                ))
-            );
+impl LogSegmentBuilder {
+    fn new(log_root: Url) -> Self {
+        LogSegmentBuilder {
+            log_root,
+            end_version: None,
         }
+    }
 
-        Ok(LogSegment {
-            end_version: effective_version,
-            checkpoint_version,
-            log_root,
-            ascending_commit_files,
-            ascending_compaction_files,
-            checkpoint_parts,
-            latest_crc_file,
-        })
+    /// Sets the end version for the log segment. This is optional, and if not set, the LogSegment
+    /// will LIST to determine the end version at the latest version of the table.
+    #[internal_api]
+    pub(crate) fn with_end_version_opt(mut self, end_version: Option<Version>) -> Self {
+        self.end_version = end_version;
+        self
```

Comment:
What other options do we expect for LogSegmentBuilder? Currently, it seems like we only ever configure a single value with the builder.